### DISCUSSION TITLE
Ratings help using title formatting (or other options)

### DISCUSSION THREAD

**Discussion by MountainX** [DeaDBeeF-Player/deadbeef#2677] [General]
### What's going on? Describe the problem in as much detail as possible.

My flac files all have a 0-100 rating stored in an tag named `rating`. I display this in DeadBeef with a custom column that has the format:

    %rating%

I was reading older issues today and I learned how to display my ratings as stars. The following works with my 0-100 rating range and translates it to 5 stars for display only:

    $progress2($max(0,$min(100,%rating%)),100,5,★,☆)

This also works and displays the numerical rating across ten stars:

    $progress2($max(0,$min(100,%rating%)),100,10,★,☆)

I'm really enjoying this. I'd like to figure out how to use half-stars in a total length of 5 stars, and maybe even round up (such as set 4 stars for a rating of 79).

However, to make it really useful, I need to be able to click on a star to set a numerical rating value that gets saved back to the track's metadata. Other music players have this functionality (Quod Libet, Strawberry, etc.). This is my main interest in regard to this issue: **how to save ratings with one click by clicking on a star.**

Saving it back to the existing `rating` tag would make the most sense.  I'm just going thru all the title formatting options, but this one caught my eye:

    $put(name,value)
    Stores value in the variable name and returns value unaltered.

Is that part of the puzzle to accomplishing this goal? Or is title formatting strictly a read-only operation?

If title formatting is not the answer, is there an existing way to accomplish the goal of a one-click method to save ratings for tracks? 

If there is no existing method, please consider this a feature request. 

I am aware that I can change a numerical rating value within deadbeef by these steps:

1. right click the track
2. select track properties
3. click the rating field's value in a somewhat finicky way to get into edit mode. This usually requires more than one click.
4. backspace over the rating. (In my experience nothing else works.)
5. type the new rating
6. click outside the field
7. click apply
8. click close to dismiss the dialog

In my experience, if I don't do all those steps very carefully, my rating change does not get saved.  It requires multiple mouse clicks (both right and left) as well as multiple typing steps.

In contrast, some other plays allow setting a rating with these steps:

1. click on the desired star value

That's it. Done. 

Ratings are one of the main ways I manage my music and it is getting very tedious to do this with the 8 steps above. I'm willing to put a bounty on this feature if it requires development.

Thank you for the great music player!

### Information about the software:

Deadbeef version: DeaDBeeF 1.8.8 
OS: Arch Linux

**Comment by Oleksiy-Yakovenko** [comment]
Try searching foobar 2000 forums, there's a whole thread on it:
https://hydrogenaud.io/index.php?PHPSESSID=9n1ssr2olf9u76sv86bu6vlbd3&topic=59187.25

I'm sure you'll find something similar to what you need.

**Comment by Oleksiy-Yakovenko** [comment]
Regarding your other questions, there's a rating plugin which you could download, which allows to set ratings using a menu. But I don't know whether it supports configuring a range to be 0..100.

> Or is title formatting strictly a read-only operation?

Title formatting is read only. It's used for converting and displaying metadata in various ways. It can't be used to modify metadata.

**Reply by MountainX** [reply]
> there's a rating plugin which you could download, which allows to set ratings using a menu. 

Thank you. Your reply is very helpful. I downloaded the song rating plugin. It is a big improvement and I will use it. 

> But I don't know whether it supports configuring a range to be 0..100.

I don't see any configuration options for that plugin, and it saves ratings as values between 0 and 5. This results in other players displaying a rating of 5 as zero stars (because they generally expect ratings like 0 to 100 or 0 to 255 or similar).

> Title formatting is read only. It's used for converting and displaying metadata in various ways. It can't be used to modify metadata.

OK, thanks.

**Reply by kpcee** [reply]
> I don't see any configuration options for that plugin, and it saves ratings as values between 0 and 5. This results in other players displaying a rating of 5 as zero stars (because they generally expect ratings like 0 to 100 or 0 to 255 or similar).

The rating plugin works internally with values from 0-5, the conversion of the values to 1-255 is done by DeaDBeeF when saving, the plugin has no influence on it. A value of 5 stars creates a Popularimeter frame with the value 254 when saving, which should also be displayed in other players with maximum stars.

**Comment by Oleksiy-Yakovenko** [comment]
@kpcee I guess it depends on file format though. I don't think there's any mapping of rating value for FLAC.

**Reply by MountainX** [reply]
> @kpcee I guess it depends on file format though. I don't think there's any mapping of rating value for FLAC.

You are correct. For FLAC it simply saves the ratings as ints from 1 to 5 (maybe 0 is included). That is not very compatible with other applications. 

All the music players I have been looking at recently save FLAC ratings as either ints 0 to 100 or decimals (like percentages) from 0.0 to 1.0. 

If this plug used either of those for FLAC it would be much better. Can I suggest this?

0 stars = 0.0
1 star  = 0.2
2 stars = 0.4
3 stars = 0.6
4 stars = 0.8
5 stars = 1.0

(or just use 0 to 100)

I think the best solution would be to add some config options for the song ratings plugin.

**Comment by Oleksiy-Yakovenko** [comment]
This would require modifying every tag read/write in deadbeef, to ensure the rating gets converted into a common range on read, and then converted into a format-specific range on write. It should not be done by the plugin.

**Reply by MountainX** [reply]
> This would require modifying every tag read/write in deadbeef, to ensure the rating gets converted into a common range on read, and then converted into a format-specific range on write. It should not be done by the plugin.

I was thinking the opposite. My library is already fully tagged (rated) using the standard range used by other apps. If I use this plugin, those ratings are lost and a range of 0-5 is saved to my file's metadata. This makes the ratings unusable in my other apps. 

It also leaves my library in an inconsistent state: if I have not used the DeadBeef plugin on songs, those songs are rated in the standard range, but if I have used the DeadBeef plugin, those songs have a non-standard rating in the range of 0-5, and this is exactly what I wish to avoid.

All I am after, and I think all most users would want, is consistency. We all know there is no formal standard for the rating metadata for FLAC. But there are two widely used informal standards: 0.0 to 1.0 as decimals, or 0 to 100 as ints. No major music app that I have seen uses 0 to 5 for rating flac files, so why make that the standard for DeadBeef by allowing this plugin to force that range?

**Comment by Oleksiy-Yakovenko** [comment]
> I was thinking the opposite. 

Not sure why you'd think the opposite. The point of what I wrote is to make ratings use different range, depending on the format, and be compatible with other apps.

**Reply by MountainX** [reply]
> > I was thinking the opposite.
> 
> Not sure why you'd think the opposite. 

Again, I apologize for poor communication. 

> The point of what I wrote is to make ratings use different range, depending on the format, and be compatible with other apps.

That would be great!

**Comment by Oleksiy-Yakovenko** [comment]
> so why make that the standard for DeadBeef by allowing this plugin to force that range?

There's no such thing. Deadbeef itself doesn't have functionality for changing the ratings in media files. The plugin does it in some way that seemed appropriate, and it works for some file formats, but not the others.

**Reply by MountainX** [reply]
> > so why make that the standard for DeadBeef by allowing this plugin to force that range?
> 
> There's no such thing. Deadbeef itself doesn't have functionality for changing the ratings in media files. The plugin does it in some way that seemed appropriate, and it works for some file formats, but not the others.

That's poor communication on my side. Sorry for the confusion. It sounds like we both agree that this plugin is not using an appropriate rating range for flac files. I don't know if there are any plans to fix it, and I know you did not create the plugin. So I'm not sure what my solution is.

I also want to add that I appreciate your work creating and maintaining DeadBeef. It is an awesome app. Thank you.

**Comment by kpcee** [comment]
I have tested this with a FLAC file with Vorbis Comments and a MP3 with ID3v2 tag. Creating a rating frame in both files with the plugin or e.g. kid3-cli, the rating frame can be viewed and changed via the internal Track Properties dialog. After changing a rating from 1 to 5 and saving it, DeaDBeeF changes the value to 255 when saving the ID3v2 tag, but in the Vorbis Comments it stays at the value 5. 

I have no problem with this behavior since I don't use FLAC or using another player. However, I can also understand MountainX point of view, since the Vorbis Comments ratings saved by DeaDBeeF are only displayed correctly under DeaDBeeF.

**Reply by MountainX** [reply]
> I have tested this with a FLAC file with Vorbis Comments and a MP3 with ID3v2 tag. Creating a rating frame in both files with the plugin or e.g. kid3-cli, the rating frame can be viewed and changed via the internal Track Properties dialog. After changing a rating from 1 to 5 and saving it, DeaDBeeF changes the value to 255 when saving the ID3v2 tag, but in the Vorbis Comments it stays at the value 5.
> 
> I have no problem with this behavior since I don't use FLAC or using another player. However, I can also understand MountainX point of view, since the Vorbis Comments ratings saved by DeaDBeeF are only displayed correctly under DeaDBeeF.

Thank you (both) for replying. Most of my music consists of FLAC files.

I'm not a developer, but I can do extensive testing and other things. I volunteer to assist if you want to change the way the deadbeef saves ratings metadata in FLAC files.

**Comment by Oleksiy-Yakovenko** [comment]
The point is that the tag reading and writing is responsibility of Deadbeef, not the plugin, and conversion of rating range belongs there.
The plugin only provides the UI pretty much.

To fix the issue with incorrect rating ranges -- Deadbeef would have to convert ratings to some common representation when reading, and then convert back to the file format when writing, regardless of the plugins.

**Reply by MountainX** [reply]
> The point is that the tag reading and writing is responsibility of Deadbeef, not the plugin, and conversion of rating range belongs there. The plugin only provides the UI pretty much. To fix the issue with incorrect rating ranges -- Deadbeef would have to convert ratings to some common representation when reading, and then convert back to the file format when writing, regardless of the plugins.

OK, thanks for explaining. I get it now.

Do you plan do implement this in the future?

As you know, there is no formal spec for rating metadata values for Vorbis/FLAC (or pretty much any other file format). However, as I have been looking into the approach different music players have adopted, it seems GMusicBrowser is doing things in a good way. Maybe you could do something compatible with that. 

In a nutshell, GMB saves ratings metadata as 0.0 to 1.0 into a tag named FMPS_RATING for Vorbis and APE. The app reads those and translates them into 0 to 100 and displays 0 to 5 stars. QuodLibet is using the same ratings scale in file metadata.

**Reply by Oleksiy-Yakovenko** [reply]
I don't have any particular plans about that, but I created an issue with proposed implementation.
This should help to align with other devs about how this should be implemented, regardless who and when decides to work on it.
https://github.com/DeaDBeeF-Player/deadbeef/issues/2681
